File: src/main/scala/net/ypmania/jio/Scope.scala
package net.ypmania.jio

class Scope (private val zioScope: zio.Scope) {
  def addFinalizer(run: UJIO[Object, ?]): UJIO[Object, Object] = {
    JIO.wrapU(zioScope.addFinalizer(JIO.unwrap(run)).as(new Object))
  }
}

object Scope {
  def make(): UJIO[Object, Scope] = JIO.wrapU(zio.Scope.make.map(s => new Scope(s)))

  trait Has {
    def scope(): Scope
  }
}

File: jio/src/main/java/net/ypmania/jio/Scope.java
package net.ypmania.jio;

import zio.Trace;


public class Scope {
    public static UJIO<Object, Scope> make() {
        return JIO.wrapU(zio.Scope.make(Trace.empty())).map(Scope::new);
    }

    private final zio.Scope zioScope;

    public Scope(zio.Scope zioScope) {
        this.zioScope = zioScope;
    }

    public UJIO<Object, Object> addFinalizer(UJIO<Object,?> run) {
        return JIO.wrapU(zioScope.addFinalizer(() -> JIO.unwrap(run.<Object>unsafeCast()), Trace.empty()));
    }

    interface Has {
        Scope scope();
    }
}
